File: web/interaction.js
import {Avatar} from "./avatar.js";
import {Data} from "./data.js";
import {Sprite} from "./sprite.js";
import {Point} from "./point.js";
import {Room} from "./room.js";

const INACTIVE = -1;
const VISIBLE = 0;
const ACTIVE = 1;
const HOVER = 2;

/** A location in a room where the avatar can trigger some sort of interaction.
 * @typedef {Object} Interaction
 */
export class Interaction {
  /** Read an interaction definition from the given data.
   * @param {Data} data
   */
  constructor(data) {
    this._name = data.value();
    // Unlike most data objects, an "interaction" may be embedded inside a room definition.
    // So, it doesn't need to be terminated by a newline; parsing will also terminate if a
    // line is encountered that is not a valid interaction attribute.
    while (data.next() && data.size()) {
      if (data.tag() === "position") {
        this._position = data.point(1);
      }
      else if (data.tag() === "offset") {
        this._iconCenter = data.point(1);
      }
      else if (data.tag() === "visible") {
        this._radius[VISIBLE] = data.point(1);
        if (data.size() >= 3) this._icon[VISIBLE] = Sprite.get(+data.arg(2));
      }
      else if (data.tag() === "active") {
        this._radius[ACTIVE] = data.point(1);
        if (data.size() >= 3) this._icon[ACTIVE] = Sprite.get(+data.arg(2));
        if (data.size() >= 4) this._icon[HOVER] = Sprite.get(+data.arg(3));
      }
      else if (data.tag() === "enter") {
        this._enterPosition = data.point(1);
        this._enterRoom = data.arg(2);
      }
      else if (data.tag() === "dialog") {
        this._dialog = data.value();
      }
      else break;
    }
    if (this._position) this._iconCenter = this._iconCenter.add(this._position);
  }

  /** Get the name of this interaction.
   * @returns {string} The interaction's name, or an empty string if no name is set.
   */
  name() {
    return this._name;
  }

  /** Set the state of the intersection. Return true if it must trigger immediately.
   * @param {Avatar} avatar
   * @returns {boolean} True if this interaction triggers immediately due to avatar movement.
   */
  setState(avatar) {
    let offset = avatar.sub(this._position);
    if (inRange(offset, this._radius[ACTIVE])) {
      if (this._state !== ACTIVE) {
        this._state = ACTIVE;
        if (!this._icon[ACTIVE]) return true;
      }
    }
    else if (this._icon[VISIBLE] && inRange(offset, this._radius[VISIBLE])) {
      this._state = VISIBLE;
    }
    else this._state = INACTIVE;
    // Return false unless we triggered an immediate action, above.
    return false;
  }

  /** Update this icon's hover state. Return true if its visible state changes.
   * @param {Point} mouse The mouse position in room coordinates.
   * @returns {boolean} True if the display needs to be updated.
   */
  setHover(mouse) {
    // Mouse hover does nothing if this interaction does not have both an active
    // icon and a hover icon.
    if (!this._icon[ACTIVE] || !this._icon[HOVER]) return false;

    const prev = this._hover;
    this._hover = this._click(mouse);
    return (this._state == ACTIVE && this._hover != prev)
  }

  /** Return true if the given mouse position is in the icon and the state is active.
   * @param {Point} mouse The mouse position in room coordinates.
   * @return {boolean} True if a click at that position activates this interaction.
   */
  click(mouse) {
    return (this._state == ACTIVE || this._state == VISIBLE) && this._click(mouse);
  }

  /** Return true if this interaction is active, i.e. clicking triggers it.
   * @return {boolean} True if the interaction is active.
   */
  isActive() {
    return this._state == ACTIVE;
  }

  /** Clear an interaction's state, e.g. due to the avatar leaving the room. */
  clearState() {
    this._state = INACTIVE;
    this._hover = false;
  }

  /** Get the interaction's position in the room.
   * @returns {Point} The position, in room coordinates.
   */
  position() {
    return this._position;
  }

  /** Get the icon to be drawn for this interaction in its current state.
   * @returns {?Sprite} The sprite to use, or null if none should be shown.
   */
  icon() {
    if (this._state == INACTIVE) return null;
    if (this._state == ACTIVE && this._hover && this._icon[HOVER]) return this._icon[HOVER];
    return this._icon[this._state];
  }

  /** Get the position where the icon should be drawn.
   * @returns {Point} The icon's center, in room coordinates.
   */
  iconCenter() {
    return this._iconCenter;
  }

  /** If this interaction triggers an "enter" event, get the point the avatar moves to.
   * @returns {?Point} The new coordinates, or null if there is no "enter" event.
   */
  enterPosition() {
    return this._enterPosition;
  }

  /** If this interaction triggers an "enter" event, get the room the avatar moves to.
   * @returns {?Room} The new room, or null if the avatar stays in the current room.
   */
  enterRoom() {
    return this._enterRoom;
  }

  /** If this interaction triggers a dialog, get the name of the dialog.
   * @returns {?string} The name of the dialog to begin.
   */
  dialog() {
    return this._dialog;
  }

  // Check if the given mouse location is inside where the active icon would be.
  _click(mouse) {
    const icon = this._icon[ACTIVE];
    if (!icon) return false;
    // Convert the mouse to an offset from the top left corner of the sprite.
    const off = this._iconCenter.sub(mouse).sub(icon.cornerOffset());
    return (off.x >= 0 && off.x < icon.width() && off.y >= 0 && off.y < icon.height());
  }

  /** @type {string} */
  _name = "";

  /** @type {Point} */
  _position = new Point();
  /** @type {Point} */
  _iconCenter = new Point();
  /** @type {Point[]} */
  _radius = [null, null];

  /** @type {number} */
  _state = INACTIVE;
  /** @type {boolean} */
  _hover = false;
  /** @type {Sprite[]} */
  _icon = [null, null, null];

  /** @type {?Point} */
  _enterPosition = null;
  /** @type {?Room} */
  _enterRoom = null;
  /** @type {?string} */
  _dialog = null;
}


// Check if the given point is within the given radius.
function inRange(point, radius) {
  // If the radius is (0, 0) it means the whole room is in range.
  if (!radius.x && !radius.y) return true;
  return (point.x / radius.x)**2 + (point.y / radius.y)**2 <= 1;
}

File: desktop/Interaction.cpp
/* Interaction.cpp
Copyright 2020 Michael Zahniser
*/

#include "Interaction.h"

#include <cstdint>

using namespace std;

namespace {
	// Check if the given point is within the given radius of the origin.
	// Note: if the radius is (0, 0), this will always return true.
	bool InRange(Point point, Point radius)
	{
		// That is, it's within a circle defined by:
		// (point.x / radius.x)^2 + (point.y / radius.y)^2 < 1
		// (point.x * radius.y)^2 + (point.y * radius.x)^2 < (radius.x * radius.y)^2
		// Use 64-bit math to avoid overflow.
		int64_t px = point.X() * radius.Y();
		int64_t py = point.Y() * radius.X();
		int64_t r = radius.X() * radius.Y();
		return (px * px + py * py <= r * r);
	}
};



// Constructor, to allow an Interaction to be emplaced.
Interaction::Interaction(Data &data)
{
	Load(data);
}



// Load an interaction definition. Usually, this will be nested inside a
// room definition.
void Interaction::Load(Data &data)
{
	name = data.Value();
	while(data.Next() && data.Size())
	{
		if(data.Tag() == "position")
			position = data[1];
		else if(data.Tag() == "offset")
			offset = data[1];
		else if(data.Tag() == "visible")
		{
			radius[VISIBLE] = data[1];
			icon[VISIBLE] = data[2];
		}
		else if(data.Tag() == "active")
		{
			radius[ACTIVE] = data[1];
			icon[ACTIVE] = data[2];
			icon[HOVER] = data[3];
			// If no hover icon is provided, use the active icon.
			if(!icon[HOVER])
				icon[HOVER] = icon[ACTIVE];
		}
		else if(data.Tag() == "enter")
		{
			hasEnter = true;
			enterPosition = data[1];
			enterRoom = data.Value(2);
		}
		else if(data.Tag() == "dialog")
			dialog = data.Value();
		// If the tag is none of the above, we have reached the end of this
		// interaction definition. The next interaction, or some other room
		// data, may begin immediately without a blank line.
		else
			break;
	}
}



// Get the name of this interaction. Names are optional; they are needed
// only for interactions that may be removed via dialog events.
const string &Interaction::Name() const
{
	return name;
}



// Set this interaction's name.
void Interaction::SetName(const string &name)
{
	this->name = name;
}



// Write this interaction to the given output stream.
void Interaction::Save(ostream &out, const string &indent) const
{
	out << indent << "interaction";
	if(!name.empty())
		out << ' ' << name;
	out << '\n';
	// Every interaction ought to have a position; everything else is optional.
	out << indent << "  position " << position.X() << ',' << position.Y() << '\n';
	if(offset)
		out << indent << "  offset " << offset.X() << ',' << offset.Y() << '\n';
	if(icon[VISIBLE])
		out << indent << "  visible " << radius[VISIBLE].X() << ',' << radius[VISIBLE].Y() << ' ' << icon[VISIBLE] << '\n';
	if(radius[ACTIVE] || icon[ACTIVE])
	{
		out << indent << "  active " << radius[ACTIVE].X() << ',' << radius[ACTIVE].Y();
		if(icon[ACTIVE])
			out << ' ' << icon[ACTIVE];
		if(icon[HOVER])
			out << ' ' << icon[HOVER];
		out << '\n';
	}
	if(hasEnter)
	{
		out << indent << "  enter " << enterPosition.X() << ',' << enterPosition.Y();
		if(!enterRoom.empty())
			out << ' ' << enterRoom;
		out << '\n';
	}
	if(!dialog.empty())
		out << indent << "  dialog " << dialog << '\n';
}



// Get the state of this interaction given the avatar's position.
int Interaction::SetState(Point avatar)
{
	// If no active radius is specified, the interaction activates whenever the
	// avatar enters this room. If no active icon is specified, the interaction
	// triggers immediately (but does not trigger again until it ceases to be 
	// active and then becomes active again).
	avatar -= position;
	if(InRange(avatar, radius[ACTIVE]))
	{
		if(state != ACTIVE)
		{
			state = ACTIVE;
			if(!icon[ACTIVE])
				return IMMEDIATE;
		}
	}
	else if(icon[VISIBLE] && InRange(avatar, radius[VISIBLE]))
		state = VISIBLE;
	else
		state = INACTIVE;
	
	return state;
}



// Manually set this interaction's state.
void Interaction::SetState(int state)
{
	this->state = state;
}



// Clear the state of this interaction. This must be done when leaving the
// room to reset any active immediate interactions.
void Interaction::ClearState()
{
	state = INACTIVE;
}



// Get the most recently set state.
int Interaction::State() const
{
	return state;
}



// Get the icon for this interaction given whatever state is set.
int Interaction::Icon() const
{
	// Check if the icon should be shown.
	if(state != VISIBLE && state != ACTIVE)
		return 0;
	
	return icon[state];
}



// Get the icon to be used when the active icon is hovered.
int Interaction::HoverIcon() const
{
	return icon[HOVER];
}



// Get the "floor" position of this interaction.
Point Interaction::Position() const
{
	return position;
}



// Get the offset of the icon relative to the sprite's floor position.
Point Interaction::Offset() const
{
	return offset;
}



// Set the interaction's position and offset.
void Interaction::Place(Point position, Point offset)
{
	this->position = position;
	this->offset = offset;
}



// Check if this interaction causes an "enter" event.
bool Interaction::HasEnter() const
{
	return hasEnter;
}



const string &Interaction::EnterRoom() const
{
	return enterRoom;
}



Point Interaction::EnterPosition() const
{
	return enterPosition;
}



// Check if this interaction causes a dialog to run.
bool Interaction::HasDialog() const
{
	return !dialog.empty();
}



const string &Interaction::DialogName() const
{
	return dialog;
}


	
// Get the radius for the given state.
const Point &Interaction::Radius(int state) const
{
	return radius[state];
}
